compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs

```diff
@@ -808,14 +801,34 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
 
         if let ObligationCauseCode::ImplDerivedObligation(obligation) = &*code {
             let expected_trait_ref = obligation.parent_trait_ref;
-            let new_imm_trait_ref = poly_trait_ref
+            let found_ty = expected_trait_ref.self_ty().skip_binder();
+            let imm_borrowed_found_ty = self.tcx.mk_imm_ref(self.tcx.lifetimes.re_static, found_ty);
+            let mut_borrowed_found_ty = self.tcx.mk_mut_ref(self.tcx.lifetimes.re_static, found_ty);
+            let imm_substs = self.tcx.mk_substs_trait(
+                imm_borrowed_found_ty,
+                &obligation.parent_trait_ref.skip_binder().substs[1..],
+            );
+            let mut_substs = self.tcx.mk_substs_trait(
+                mut_borrowed_found_ty,
+                &obligation.parent_trait_ref.skip_binder().substs[1..],
+            );
```

Comment:
Looks like a common logic here and below can be extracted into a closure (or function).

Reply:
There's more common logic than the new commit extracts, the closure can take `found_ty` and `trait_ref` and return the `(imm_substs, mut_substs)` pair.

(Could you also squash commits after addressing this comment?)

Reply:
Ah, I see what you mean. Let me know if this version is what you intended.